cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -1,55 +1,247 @@
 mod cache;
-mod transaction_pattern;
 mod web3_connector;
 
-pub use self::{
-    cache::Cache,
-    transaction_pattern::{Event, Topic, TransactionPattern, TRANSACTION_STATUS_OK},
-    web3_connector::Web3Connector,
-};
+pub use self::{cache::Cache, web3_connector::Web3Connector};
 use crate::{
     btsieve::{BlockByHash, LatestBlock, Predates, ReceiptByHash},
-    ethereum::{Transaction, TransactionAndReceipt, TransactionReceipt, H256, U256},
+    ethereum::{Address, Bytes, Transaction, TransactionReceipt, H256, U256},
     Never,
 };
 use anyhow;
 use chrono::NaiveDateTime;
-use futures_core::compat::Future01CompatExt;
+use ethbloom::Input;
+use futures_core::{compat::Future01CompatExt, Future};
 use genawaiter::{
     sync::{Co, Gen},
     GeneratorState,
 };
-use std::collections::HashSet;
+use std::{collections::HashSet, pin::Pin};
 
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub const TRANSACTION_STATUS_OK: u32 = 1;
+
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<(Transaction, Address)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, receipt) =
+        matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+            // transaction.to address is None if, and only if, the transaction
+            // creates a contract.
+            transaction.to.is_none() && transaction.input == bytecode
+        })
+        .await?;
+
+    match receipt.contract_address {
+        Some(location) => Ok((transaction, location)),
+        None => return Err(anyhow::anyhow!("contract address missing from receipt")),
+    }
+}
+
+pub async fn matching_event_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    event: Event,
+    action: &'static str,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_bloom_filter(
+        blockchain_connector.clone(),
+        start_of_swap,
+        event.topics.clone(),
```

Comment:
With the matcher being non-async, this could turn into a reference.

Reply:
Had a go but the `event` being used in the closure as well as here means we still need to clone.

Reply:
Only if the closure `move`s it right? If it is not async, there shouldn't be a need to move it. My brain borrow checker is weak though, so could very easily be that I am wrong :D